File: python/65_shower_line_431B.py
import itertools
import sys

g = [list(map(int, sys.stdin.readline().split())) for _ in range(5)]
best = 0
for perm in itertools.permutations(range(5)):
    res = 0
    q = list(perm)
    while q:
        for i in range(0, len(q) // 2 * 2, 2):
            u, v = q[i], q[i + 1]
            res += g[u][v] + g[v][u]
        q.pop(0)
    best = max(best, res)
print(best)

File: cpp/65_shower_line_431B.cpp
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int g[5][5];
    for (int i = 0; i < 5; i++)
        for (int j = 0; j < 5; j++)
            cin >> g[i][j];
    vector<int> p = {0, 1, 2, 3, 4};
    int best = 0;
    do
    {
        int res = 0;
        vector<int> q = p;
        int now = 5;
        while (now > 0)
        {
            int paired = now / 2;
            for (int i = 0; i < paired; i++)
            {
                int u = q[2 * i], v = q[2 * i + 1];
                res += g[u][v] + g[v][u];
            }
            q.erase(q.begin());
            now--;
        }
        best = max(best, res);
    } while (next_permutation(p.begin(), p.end()));
    cout << best << "\n";
}
